src/provider/provider_os_memory.c

```diff
@@ -83,28 +87,64 @@ static void os_store_last_native_error(int32_t native_error, int errno_value) {
 
 static umf_result_t nodemask_to_hwloc_nodeset(const unsigned *nodelist,
                                               unsigned long listsize,
-                                              hwloc_bitmap_t *out_nodeset) {
-    if (out_nodeset == NULL) {
-        return UMF_RESULT_ERROR_INVALID_ARGUMENT;
-    }
+                                              os_memory_provider_t *os_provider,
+                                              int is_separate) {
 
-    *out_nodeset = hwloc_bitmap_alloc();
-    if (!*out_nodeset) {
+    unsigned long array_size = listsize && is_separate ? listsize : 1;
+    os_provider->nodeset =
+        umf_ba_global_alloc(sizeof(*os_provider->nodeset) * array_size);
+
+    if (!os_provider->nodeset) {
         return UMF_RESULT_ERROR_OUT_OF_HOST_MEMORY;
     }
 
+    hwloc_bitmap_t *out_nodeset = os_provider->nodeset;
+    os_provider->nodeset_len = array_size;
     if (listsize == 0) {
+        // Hwloc_set_area_membind fails if empty nodeset is passed so
```

Comment:
This is an internal function so we should strictly control how we call it. We should assert here.

Reply:
i do not know what do you want to assert here